numba/smartarray.py

```diff
@@ -0,0 +1,212 @@
+from numba.tracing import trace
+import numpy
+import sys
+
+def _o2s(dtype, shape, order):
+    # convert order parameter to strides
+
+    if dtype is None or shape is None or order is None:
+        return None
+
+    if order == 'F':
+        shape = list(shape)
+        shape.reverse()
+    strides = []
+    itemsize = dtype.itemsize
+    for i in range(len(shape), 0, -1):
+        strides.append(itemsize)
+        itemsize *= shape[i - 1]
+    if order in ('C', None):
+        strides.reverse()
+    return tuple(strides)
+
+def _s2o(dtype, shape, strides):
+    # convert strides parameter to order
+    # Note: strides must correspond to contiguous data layout
+
+    if strides is None or strides[-1] == dtype.itemsize:
+        order = 'C'
+    elif strides[0] == dtype.itemsize:
+        order = 'F'
+    else:
+        raise ValueError('strides do not correspond to contiguous data layout')
+    s2 = _o2s(dtype, shape, order)
+    if strides != s2:
+        raise ValueError('strides do not correspond to contiguous data layout')
+    return order
+
+class SmartArray(object):
+    """An array type that supports host and GPU storage."""
+
+    def __init__(self, obj=None, copy=True,
+                 shape=None, dtype=None, order=None, where='host'):
+        """Construct a SmartArray in the memory space defined by 'where'.
+        Valid invocations:
+
+        * SmartArray(obj=<array-like object>, copy=<optional-true-or-false>):
+
+          to create a SmartArray from an existing array-like object.
+          The 'copy' argument specifies whether to adopt or to copy it.
+
+        * SmartArray(shape=<shape>, dtype=<dtype>, order=<order>)
+
+          to create a new SmartArray from scratch, given the typical NumPy array
+          attributes.
+
+        (The optional 'where' argument specifies where to allocate the array
+        initially. (Default: 'host')
+        """
+
+        assert where in ('host', 'gpu')
+        # we need either a prototype or proper type info
+        assert obj is not None or (shape and dtype)
+        self._host = self._gpu = None
+        self._host_valid = self._gpu_valid = False
+        self._allocate(where, obj, dtype, shape, _o2s(dtype, shape, order), copy)
+        if where == 'host':
+            self._host_valid = True
+            t = self._host
+        else:
+            self._gpu_valid = True
+            t = self._gpu
+        self._shape = t.shape
+        self._strides = t.strides
+        self._dtype = t.dtype
+        self._ndim = t.ndim
+        self._size = t.size
+
+    @property
+    def shape(self): return self._shape
+
+    @property
+    def strides(self): return self._strides
+
+    @property
+    def dtype(self): return self._dtype
+
+    @property
+    def ndim(self): return self._ndim
+
+    @property
+    def size(self): return self._size
+
+    def host(self):
+        """Return the host representation of 'self'."""
+
```

Comment:
There probably is (if nothing but habit) :-) I think I find this a little more readable that way. But I don't have a strong opinion.